### ISSUE TITLE
Interoperation with windowing libraries

### ISSUE THREAD

**Issue by kryptan** [webgpu-native/webgpu-headers#149]
Right now creating surface [requires platform specific code](https://github.com/gfx-rs/wgpu-native/blob/master/examples/triangle/main.c#L58). This is unfortunate.

Let's say I'm writing a cross platform windowing library (similar to GLFW and SDL) and want to add convenient cross platform API to assist with creation of webgpu-native surfaces. I don't want to add dependency on any particular WebGPU implementation, just to provide an API such that if user decides to use webgpu-native then there is no need to write platform specific code.

Would such an API look like this (assuming a patch to GLFW)?

```C
// returns one of WGPUSurfaceDescriptorFrom*
WGPUChainedStruct* glfwGetWebGPUSurfaceDescriptor(GLFWwindow *window);

// release memory allocated for descriptor
void glfwFreeWebGPUSurfaceDescriptor(WGPUChainedStruct*);
```

This looks reasonable but just want to make sure that this is the API that windowing libraries are supposed to implement.

**Comment by kvark** [comment]
This seems like a viable path eventually, but I wouldn't expect GLFW, SDL, or other libraries to do anything on this front until webgpu C header is stable.

The other path is to come up with something like Rust's [raw-windowhandle](https://github.com/rust-windowing/raw-window-handle). This would mean the following steps:
  1. define a header-style library with the only purpose if storing a window handle
  2. implement support for this library (consuming the data) in webgpu-headers (this repo)
  3. implement support for this library (producing the data) for every multimedia library that needs to work with WebGPU, i.e. GLFW, SDL, and others
  4. Profit!

This approach is more involved, but it has a few benefits. The minor one is that it's easier to convince other library providers to integrate, because producing a raw window handle may be useful for other graphics layers, not just WebGPU. The major benefit is that this can be done here and now, even without webgpu-headers being stable.

**Comment by Kangz** [comment]
If adding a C version of `raw-windowhandle` is too involved, you could also add a small utility function like Dawn's utility [`CreateSurfaceFromWindow`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/utils/GLFWUtils.h;l=33;bpv=0;bpt=0) that consumes a GLFW window and gives you  a `WGPUSurface`. It is quite simple (except the small annoyance with needed and Objective C file for Metal). (please ignore the GL part, Dawn's GL backend is still WIP)

**Comment by pplux** [comment]
To be honest that kind of approach is not that uncommon and the fact that GFLW exposes the native handles it is precisely for this particular purpose. The code itself is not that much

Given a "nativeHandle" ... just a pointer, this is the platform dependent code that glues that handle with the surface descriptor initialization. 

```cpp
    WGPUSurfaceDescriptor surfaceDescriptor = {};
#if PLATFORM == PLATFORM_WINDOWS
    WGPUSurfaceDescriptorFromWindowsHWND surfaceDescriptorWin = {};
    surfaceDescriptorWin.chain.sType = WGPUSType_SurfaceDescriptorFromWindowsHWND;
    surfaceDescriptorWin.hinstance = GetModuleHandle(nullptr);
    surfaceDescriptorWin.hwnd = nativeHandle;
    surfaceDescriptor.nextInChain = (const WGPUChainedStruct*) &surfaceDescriptorWin;
#elif PLATFORM == PLATFORM_MAC
    WGPUSurfaceDescriptorFromMetalLayer surfaceDescriptorMac = {};
    surfaceDescriptorMac.chain.sType = WGPUSType_SurfaceDescriptorFromMetalLayer;
    surfaceDescriptorMac.layer = nativeHandle;
    surfaceDescriptor.nextInChain = (const WGPUChainedStruct*) &surfaceDescriptorMac;
#elif PLATFORM == PLATFORM_WASM
    WGPUSurfaceDescriptorFromCanvasHTMLSelector surfaceDescriptorWasm = {};
    surfaceDescriptorWasm.chain.sType = WGPUSType_SurfaceDescriptorFromCanvasHTMLSelector;
    surfaceDescriptorWasm.selector = "#canvas" ;
    surfaceDescriptor.nextInChain = (const WGPUChainedStruct*) &surfaceDescriptorWasm;
#endif
```

There is no longer GLFW involved, and this approach should work with any other windowing libraries out there, as long as they can provide with that native handle.

This is the other side of the code that extracts the handle from glfw to be used there:

```cpp

 #if PLATFORM == PLATFORM_WINDOWS
    handle = glfwGetWin32Window(glfwWinHandle);
 #elif PLATFORM == PLATFORM_MAC
    {
        id metal_layer = NULL;
        NSWindow *ns_window = glfwGetCocoaWindow(glfwWinHandle);
        [ns_window.contentView setWantsLayer:YES];
        metal_layer = [CAMetalLayer layer];
        [ns_window.contentView setLayer:metal_layer];
        initializeWGPU(metal_layer, win);
    }
 #elif PLATFORM == PLATFORM_WASM
    handle  = nullptr;
 #endif

```

This is the only part that it is glfw-specific, the first one, it is just the handle needed for webgpu and it is not specific to glfw.

**Comment by kainino0x** [comment]
webgpu.h meeting today: Labeling this non-breaking, because IIUC, this won't block anyone from doing anything, but we could support this later. The recommended path can be to use platform-specific code for now (even though it's annoying).

If and when glfw/sdl/etc. start gaining support for webgpu.h, perhaps we'd prefer if people did something like raw-windowhandle instead (if possible - I don't really know), but it's also OK if whoever ends up contributing support to those libraries just does something custom for webgpu.h.

**Comment by Kangz** [comment]
I would like to close this as "Won't fix": it is something that should be done in GLFW / SDL and other libraries, and even if they don't end up adding support for it, [dawn_glfw](https://source.chromium.org/chromium/chromium/src/+/main:third_party/dawn/src/dawn/glfw/?q=dawn%2Fglfw&ss=chromium%2Fchromium%2Fsrc) shows that it is fairly trivial to do a small wrapper. Also there is always going to be some OS-specific dispatch in places because the window types are just different.

**Comment by eliemichel** [comment]
FTR a PR has been pending on GLFW for a while: https://github.com/glfw/glfw/pull/2333

It is available as a standalone repo to be used with glfw: [glfw3webgpu](https://github.com/eliemichel/glfw3webgpu), and a similar one for SDL: [sdl2webgpu](https://github.com/eliemichel/sdl2webgpu) (which I should update, following the recent renaming of STypes)

As mentioned by Kangz Dawn provides its own version, and wgpu-native shows a similar solution in [its examples](https://github.com/gfx-rs/wgpu-native/blob/trunk/examples/triangle/main.c).
